collect/central_collector.go

```diff
@@ -119,18 +153,235 @@ func (c *CentralCollector) collect() {
 
 }
 
-func (c *CentralCollector) processSpan(sp *types.Span) {
-	// add the span to the cache
-	trace := c.SpanCache.Get(sp.TraceID)
-	if trace == nil {
-		err := c.SpanCache.Set(sp)
-		if err != nil {
-			c.Logger.Error().Logf("error adding span with trace ID %s to cache: %s", sp.TraceID, err)
-			return
+func (c *CentralCollector) processTraces() error {
+	for {
+		select {
+		case <-c.done:
+			return nil
+		default:
+			ids := c.SpanCache.GetTraceIDs(processTracesBatchSize)
+			if len(ids) == 0 {
+				continue
+			}
+
+			statuses, err := c.Store.GetStatusForTraces(context.Background(), ids)
+			if err != nil {
+				c.Logger.Error().Logf("error getting statuses for traces: %s", err)
+				continue
+			}
+			for _, status := range statuses {
+				switch status.State {
+				case centralstore.DecisionKeep:
+					c.send(status)
+					c.SpanCache.Remove(status.TraceID)
+
+				case centralstore.DecisionDrop:
+					c.SpanCache.Remove(status.TraceID)
+				default:
+					c.Logger.Debug().Logf("trace %s is still pending", status.TraceID)
+				}
+			}
+		}
+
+		select {
+		case <-c.done:
+			return nil
+		default:
+		}
+
+		ticker := c.Clock.NewTicker(processTracesBackoffInterval)
+		select {
+		case <-c.done:
+			return nil
+		case <-ticker.Chan():
+			ticker.Stop()
+			continue
 		}
 	}
+}
 
-	trace = c.SpanCache.Get(sp.TraceID)
+func (c *CentralCollector) decide() error {
+	for {
+		select {
+		case <-c.done:
+			return nil
+		default:
+			if c.BlockOnDecider {
+				continue
+			}
+			ctx := context.Background()
+			tracesIDs, err := c.Store.GetTracesNeedingDecision(ctx, deciderBatchSize)
+			if err != nil {
+				c.Logger.Error().Logf("error getting traces needing decision: %s", err)
+				continue
+			}
+
+			if len(tracesIDs) == 0 {
+				continue
+			}
+			statuses, err := c.Store.GetStatusForTraces(ctx, tracesIDs)
+			if err != nil {
+				c.Logger.Error().Logf("error getting statuses for traces: %s", err)
+				continue
+			}
+			traces := make([]*centralstore.CentralTrace, len(statuses))
+			stateMap := make(map[string]*centralstore.CentralTraceStatus, len(statuses))
+			var wg sync.WaitGroup
+			for i, status := range statuses {
+				// make a decision on each trace
+				if status.State != centralstore.AwaitingDecision {
+					// someone else got to it first
+					continue
+				}
+				stateMap[status.TraceID] = status
+
+				wg.Add(1)
+				go func(status *centralstore.CentralTraceStatus, idx int) {
```

Comment:
wondering if we want to use the errgroup functionality here too, so we can SetLimit the number of goroutines we start up here. That could also be a configurable value to keep from hammering on redis too hard. It's close to, but not identical, to `deciderBatchSize`. We probably only need one of them, but this is an alternative way to keep things under control.